execution/block-partitioner/src/pre_partition/connected_component/mod.rs

```diff
@@ -0,0 +1,161 @@
+            let sender_idx = state.sender_idx(txn_idx0);
+            let write_set = state.write_sets[txn_idx0].read().unwrap();
+            let read_set = state.read_sets[txn_idx0].read().unwrap();
+            for &key_idx in write_set.iter().chain(read_set.iter()) {
+                let key_idx_in_uf = num_senders + key_idx;
+                uf.union(key_idx_in_uf, sender_idx);
+            }
+        }
+        // NOTE: union-find result is NOT deterministic. But the following step can fix it.
+
+        // Entities & relations involved in the following processing.
+        //
+        // txn-0 txn-7 txn-9     txn-1 txn-2 txn-3 txn-4 txn-5 txn-6 txn-8
+        //      \  |  /               \    \   |     |     |   /    /
+        //       \ | /                  \   |  |     |     |  |  /
+        // conflicting-set-0                conflicting-set-1
+        //      /      \                  /  |         |     \
+        //     /        \               /    |         |      \
+        // txn-grp-0 txn-grp-1  txn-grp-2 txn-grp-3 txn-grp-4 txn-grp-5
+        //         \        \         \  /          /         /
+        //          \        \         \/          /       /
+        //            \       \        /\         /     /
+        //               \     \     /    \      /  /
+        //                  Shard-0         Shard-1
+
+        // Prepare `txns_by_set`: a mapping from a conflicting set to its txns.
+        let mut txns_by_set: Vec<VecDeque<TxnIdx0>> = Vec::new();
+        let mut set_idx_registry: HashMap<usize, usize> = HashMap::new();
+        let set_idx_counter = AtomicUsize::new(0);
+        for txn_idx0 in 0..state.num_txns() {
+            let sender_idx = state.sender_idx(txn_idx0);
+            let uf_set_idx = uf.find(sender_idx);
+            let set_idx = set_idx_registry.entry(uf_set_idx).or_insert_with(|| {
+                txns_by_set.push(VecDeque::new());
+                set_idx_counter.fetch_add(1, Ordering::SeqCst)
+            });
+            txns_by_set[*set_idx].push_back(txn_idx0);
+        }
+
+        // Calculate txn group size limit.
+        let group_size_limit = ((state.num_txns() as f32) * self.load_imbalance_tolerance
+            / (state.num_executor_shards as f32))
+            .ceil() as usize;
+
+        // Prepare `group_metadata`, a group_metadata (i, r) will later be converted to a real group that takes `r` txns from set `i`.
+        // NOTE: If we create actual txn groups now and then do load-balanced scheduling, we break the relative order of txns from the same sender.
+        // The workaround is to only fix the group set and their sizes for now, then schedule, and materialize the txn groups at the very end (when assigning groups to shards).
+        let group_metadata: Vec<(usize, usize)> = txns_by_set
+            .iter()
+            .enumerate()
+            .flat_map(|(set_idx, txns)| {
+                let num_big_chunks = txns.len() / group_size_limit;
```

Comment:
I would have named it as 'num_chunks' = txns.len() / group_size_limit + 1
Instead of 'reminder_chunk' use 'last_chunk'

Reply:
fixed